### PR TITLE
FIX: json backend atomic writes

### LINKED ISSUE #301: BUG: Interrupt during save operation can corrupt database
## Expected Behavior
Saving should always at least return the database to a readable state, but preferably complete any atomic operation before terminating.

## Current Behavior
If a save operation is interrupted, the current save operation will not be completed.  If the interruption is caught and `save()` is called again, the database can be restored. (this is the solution used in #300, but should be replaced when a more general solution is found)

The below seems to work:
```python
try:
    result.item.save()
except KeyboardInterrupt:
    result.item.save()
    break
```

## Possible Solution
Context managers that catch all exceptions and properly close out the files?  In any case we need to make the save operation safer, and more tolerant of interruption.  It's possible that something like the above try-except inside the backend `.save()` method could work, but that's speculation on my part

## Steps to Reproduce (for bugs)
Start a sequence of save actions using the happi client, even without edits to an item (though last edit date will change)
Ctrl+C / KeyboardInterrupt / SigInt 
Depending on timing the database may be in various states of completeness.  

image

Here happi repair is run on all the devices with debug output. The keyboard interrupt occurs after the backend `.update()` method appears to resolve, but it's clear the file writing is not complete.  The diff in the json database is shown on the left. 
 Changes at the top are all valid last_edited modifications, while the bottom of the database is arbitrarily truncated.  

## Context
First uncovered when working on `happi repair` #300, but has existed for a while. 

## Your Environment
pcds-5.6.0

### PR DESCRIPTION
**Pull request by ZLLentz**
## Description

Use the operating system's copy/rename/move mechanisms to implement atomic updates of the json backend.
Accordingly, the file locking has been removed, it should no longer be necessary.

Also remove some references to Travis CI that I noticed while filling in the template.

## Motivation and Context

closes #301 
Summary: if you press ctrl+c at the wrong time you can truncate your database if it uses the json backend.

## How Has This Been Tested?

I added some unit tests and the old tests still pass.
Interactively with a test db, I am able to ctrl+c and see my temp file be partially created without corrupting the main data file.

I want to make a more direct unit test that gives me a high confidence that interrupting this function at any time won't corrupt the database, but I haven't figured out a good way to interrupt the function at specific timings.

## Where Has This Been Documented?

Pre-release notes

## Pre-merge checklist
- [x] Code works interactively
- [x] Code contains descriptive docstrings, including context and API
- [x] New/changed functions and methods are covered in the test suite where possible
- [x] Test suite passes locally
- [x] Test suite passes on continuous integration ~(Travis CI)~
- [x] Ran ``docs/pre-release-notes.sh`` and created a pre-release documentation page
- [x] Pre-release docs include context, functional descriptions, and contributors as appropriate

### CODE REVIEW

**Comment by tangkong** [comment]
I'll give this a spin tomorrow, it looks pretty solid to me otherwise.  

RE interrupting the save action at specific points, the only thing I can think of is monkeypatching specific methods to throw an Exception or something inside the methods used by `JSONBackend.store` (maybe `json.dump`, `shutil.copymode`, etc)

**Review by tangkong — APPROVED** [review]
Seems to work as advertised 👍 I suppose the temp files should only appear if a write is canceled, so cleaning them up isn't necessary nor desired.  We can just periodically clean them up manually  (or add it to the cron? 🤔 )

**Comment by ZLLentz** [comment]
Yeah, I don't think the temp files should be auto-cleared in the event of failure, though we could auto-clear them some other time. I think they have positive side effects, such as letting us recover an update when the user doesn't have write permissions on the main database file.

> RE interrupting the save action at specific points: the only thing I can think of is monkeypatching specific methods to throw an Exception or something inside the methods used by JSONBackend.store (maybe json.dump, shutil.copymode, etc)

I can do this, what I really want to do though is to raise an exception during the execution of these functions rather than instead of them.

**Review by klauer — APPROVED** [review]

File: happi/backends/json_db.py [hunk 1]
```diff
@@ -128,29 +124,45 @@ def store(self, db: dict[str, ItemMeta]) -> None:
...
-            # Create lock in filesystem
-            if fcntl is not None:
-                fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
-            # Dump to file
-            try:
-                json.dump(db, f, sort_keys=True, indent=4)
+        with open(temp_path, 'w') as fd:
+            json.dump(db, fd, sort_keys=True, indent=4)
+
+        if os.path.exists(self.path):
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:145]: I think there may be a difference between the temporary file being cleaned up in the event that it doesn't get finished (is truncated, therefore useless) and in the event that the move operation fails (is complete, and therefore could be manually renamed with fixed permissions)

File: happi/backends/json_db.py [hunk 2]
```diff
@@ -128,29 +124,45 @@ def store(self, db: dict[str, ItemMeta]) -> None:
...
+            shutil.copymode(self.path, temp_path)
+        shutil.move(temp_path, self.path)
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:147]: Does that mean we could do useful things like log the underlying file operations? Or maybe more sinister things, like rig the test suite to raise during a shutil.move?

File: happi/backends/json_db.py [hunk 3]
```diff
@@ -17,14 +21,6 @@
 
 logger = logging.getLogger(__name__)
 
-
-try:
-    import fcntl
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:22]: When I realized it didn't have a purpose here any more I was very happy, it was originally just stopping us from writing out json files on top of each other

File: happi/backends/json_db.py [hunk 4]
```diff
@@ -128,29 +124,45 @@ def store(self, db: dict[str, ItemMeta]) -> None:
...
+    def _temp_path(self) -> str:
+        """
+        Return a temporary path to write the json file to during "store".
+
+        Includes a username and a timestamp for sorting
+        (in the cases where the temp file isn't cleaned up)
+        and a hash for uniqueness
+        (in the cases where multiple temp files are written at once).
+        """
+        directory = os.path.dirname(self.path)
```
[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:158]: I think you're right, there won't be enough of these files that the extra bits will be particularly helpful (files have owners and create dates after all), I'm not particularly motivated to rewrite

File: happi/backends/json_db.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (klauer) [reply happi/backends/json_db.py:147]: Yeah, interestingly that appears to be the case:
> When an auditing event is raised through the [sys.audit()](https://docs.python.org/3/library/sys.html#sys.audit) function, each hook will be called in the order it was added with the event name and the tuple of arguments. Native hooks added by [PySys_AddAuditHook()](https://docs.python.org/3/c-api/sys.html#c.PySys_AddAuditHook) are called first, followed by hooks added in the current (sub)interpreter. Hooks can then log the event, raise an exception to abort the operation, or terminate the process entirely.

[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:147]: I'm going to investigate this at least a little bit to see if I can force the method to get interrupted at various spots

[the remark this answers is not in the document]
Comment (ZLLentz) [reply happi/backends/json_db.py:147]: This is a fairly useful logging/debug tool, probably not for the test suite case that I was intending though.
```
In [1]: import sys, shutil

In [2]: def audit(event, args):
   ...:     if event.startswith('os.') or event.startswith('shutil.'):
   ...:         print(event, args)
   ...:

In [3]: sys.addaudithook(audit)

os.scandir ('.',)
os.scandir ('.',)
In [4]: shutil.move('deps.log', 'deps.txt')
shutil.move ('deps.log', 'deps.txt')
os.rename ('deps.log', 'deps.txt', -1, -1)
Out[4]: 'deps.txt'
```
the random os.scandir calls are from me invoking tab completion

**Comment by ZLLentz** [comment]
Ok maybe any other changes can be handled in a follow-up

**Pull request merged**